Adds FAQ to project readme (#27)
openpubkey/openpubkey: README.md

# OpenPubkey Reference Implementation

This repo contains the current reference implementation of OpenPubkey. The reference implementation is a work in progress.

## What is OpenPubkey?

OpenPubkey adds user or workload generated public keys to OpenID Connect (OIDC) enabling identities to sign messages or artifacts under their OIDC identity. In essence, OpenPubkey is a protocol for getting OpenID Providers (OPs) to bind identities to public keys.

Verifiers can check that these signatures are valid and associated with the signing OpenID identity. OpenPubkey does not add any new trusted parties beyond what is required for OpenID Connect. It is fully compatible with existing OpenID Providers (Google, Azure/Microsoft, Okta, OneLogin, Keycloak) without any changes to the OpenID Provider.

### ID Tokens and PK Tokens

> OpenPubkey uses ID Tokens issued by OpenID Providers (OPs) to produce PK Tokens. A PK Token consist of the ID Token and the CIC (Client-Instance Claims) where the CIC contains the user's public key and associated metadata. The ID Token contains a hash of the CIC, so that the values in the CIC including the identities public key are cryptographically bound to the ID Token signed by the OP.

Reviewer: "identities" -> "identity's"

Reviewer 2: @EthanHeilman can this be fixed?

Author: @whalelines Vscode was telling me the file was saved with this being fixed and git diff was saying no differences. Took me about twenty minutes to figure this out, but github disagreeing. Turns out vscode let me open two windows to same file, which triggered a bug in vscode so that it got confused about the actual content of the file. Had to close vscode to fix it. I was starting to question my sanity

Should be fixed now.

For workload-identity, the hash of the CIC is stored in the ID Token's `aud` claim following the pattern of GitHub Action's workload identity. For user-identity, the hash of the CIC is stored in the ID Token's `nonce` claim.

## OpenPubkey FAQ

### What is the difference between Sigstore and OpenPubkey?

OpenPubkey cannot really be compared to Sigstore. Sigstore is an end-to-end artifact signing solution, whereas OpenPubkey only binds public keys to OIDC identities for use as part of a larger signing solution. OpenPubkey is complementary to Sigstore, for example https://github.com/sigstore/fulcio/issues/1056

As stated in the Related Work section of [the OpenPubkey paper](https://eprint.iacr.org/2023/296.pdf):

> > “Sigstore [31, 42] is an open source project for signing and verifying software artifacts. Users can sign under their OpenID Connect identity by using the sigstore Fulcio Certificate Authority [43] which uses an immutable log to store a mapping between an OpenID Connect ID Token and a short lived public key enabling parties to attribute signatures to identities. The Fulcio CA (Certificate Authority) is trusted to create this mapping between an ID Token and a public key. Using OpenPubkey this trust can be eliminated as OpenPubkey does not need a trusted party to map ID Tokens to public keys. The Fulcio CA could in turn help OpenPubkey by acting as a public OpenPubkey verifier and OP public key database."

Reviewer 2: ```suggestion
> Sigstore [31, 42] is an open source project for signing and verifying software artifacts. Users can sign under their OpenID Connect identity by using the sigstore Fulcio Certificate Authority [43] which uses an immutable log to store a mapping between an OpenID Connect ID Token and a short lived public key enabling parties to attribute signatures to identities. The Fulcio CA (Certificate Authority) is trusted to create this mapping between an ID Token and a public key. Using OpenPubkey this trust can be eliminated as OpenPubkey does not need a trusted party to map ID Tokens to public keys. The Fulcio CA could in turn help OpenPubkey by acting as a public OpenPubkey verifier and OP public key database.
```

No need to quote a blockquote

### How does OpenPubkey ensure the nonce claim functions as nonce?

> In the user-identity scenario the CIC that contains the user's public key is hashed to the `nonce` claim in the ID Token. As OIDC requires that this field never repeat, OpenPubkey includes a random value, rz, in the CIC. Thus the hash of the CIC is always different and random. This maintains the required properties needed by the `nonce` claim in OIDC.

Reviewer: It most sections, the first time you mention OP you spell it out, I think that's worth doing with the CIC in this section as well

Reviewer 2: Can this be updated?

```golang
rz = crypto.random()
CIC = {'rz': crypto.random(), 'upk': <publickey>, 'alg': 'EC256'}
IDToken.nonce = SHA3(CIC)

> ```

Reviewer: just some minor adjustments for consistency's sake, it should look more like this:
```golang
cic := {
    "rz": crypto.random(), 
    "upk": <publickey>, 
    "alg": "EC256"
}

idToken := {
    ...
    "nonce": SHA3(cic),
    ...
}
```

### Does OpenPubkey present a privacy leak?

The PK Tokens used by OpenPubkey contain the claims from the OIDC ID Tokens of the signer, so making them public necessarily makes those claims public too. This may include elements of the signer’s identity such as the signer’s name or email addresses. It is up to users of OpenPubkey as to whether or not PK Tokens are made public.

In a public artifact signing scenario, it could be argued that these very claims are the very claims upon which trust in the artifact should be based. However, some OIDC providers may include claims that the signer may wish to keep private. Users of OpenPubkey should consider carefully which OIDC providers to integrate with.

> Most OpenID Providers (OP) allow you to scope the fields in the ID Tokens. For instance Google’s OP is by default scoped to only include userinfo-email claims: name, email address and icon. Given that purpose of OpenPubkey is to enable parties to verify that a particular identity e.g., ethan@bastion.com, produced a particular signature, if you do not want signatures to be associated with OIDC identities, then OpenPubkey may not be a good fit for your use case.

Reviewer: "Given that purpose" -> "Given that the purpose"

> Most OpenID Providers (OP) allow you to scope the fields in the ID Tokens. For instance Google’s OP is by default scoped to only include userinfo-email claims: name, email address and icon. Given that purpose of OpenPubkey is to enable parties to verify that a particular identity e.g., ethan@bastion.com, produced a particular signature, if you do not want signatures to be associated with OIDC identities, then OpenPubkey may not be a good fit for your use case.

Reviewer: ```suggestion
Most OpenID Providers (OP) allow you to scope the fields in the ID Tokens. For instance Google’s OP is by default scoped to only include userinfo-email claims: name, email address and icon. Given that the purpose of OpenPubkey is to enable parties to verify that a particular identity (e.g. ethan@bastion.com) produced a particular signature, if you do not want signatures to be associated with OIDC identities, then OpenPubkey may not be a good fit for your use case.
```

Author: You know I love parenthetical remarks but I prefer "e.g., <instance>," over "(e.g., <instance>)" likely due to the influence of Sharon. Both are correct.

I like your other reword

In other OpenPubkey deployment scenarios, such as those employed by BastionZero, the ID Tokens are not made publicly available.

### Can the ID Tokens contained in PK Token be replayed against OIDC Resource Providers?

> Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OP's (OpenID Provider) signature can be stripped from the ID Token, and a proof of the OP's signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been stripped. However the GQ Signature provides the same assurance that the original OP's signature did.

Reviewer: ```suggestion
Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OpenID Provider's (OP) signature can be stripped from the ID Token, and a zero-knowledge proof of the OP's signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been removed without compromising any of the assurances that the original OP's signature provided.
```

Reviewer: maybe we should provide a link to an explainer about GQ?

Author: We should be we don't have a good explainer yet

We follow the approach specified in the paper: [Reducing Trust in Automated Certificate Authorities via Proofs-of-Authentication.](https://arxiv.org/abs/2307.08201)

For user-identity scenarios in which the PK Token is not made public, GQ signatures are not required. Current GQ Signatures are required for all workload-identity usecases.

### Is it a problem that GQ Signatures only work with RSA signatures?

No because the OpenID Connect spec requires that all OPs (OpenID Providers) support RSA signatures.

> > "OPs MUST support signing ID Tokens with the RSA SHA-256 algorithm (an alg value of RS256), unless the OP only supports returning ID Tokens from the Token Endpoint (as is the case for the Authorization Code Flow) and only allows Clients to register specifying none as the requested ID Token signing algorithm."

Reviewer 2: ```suggestion
> OPs MUST support signing ID Tokens with the RSA SHA-256 algorithm (an alg value of RS256), unless the OP only supports returning ID Tokens from the Token Endpoint (as is the case for the Authorization Code Flow) and only allows Clients to register specifying none as the requested ID Token signing algorithm.
```

No need to quote a blockquote

From [OpenID Connect Core 1.0 - Section 15.1 Mandatory to Implement Features for All OpenID Providers](https://openid.net/specs/openid-connect-core-1_0.html#ServerMTI).

### What about key management?

OpenPubkey assumes that all identity-held key pairs are ephemeral. You generate them as needed and delete them when you are done. No key management headaches.

### Should I be putting so much trust in a single OpenID Provider (OP)?

> Currently anyone using OIDC is trusting a single OpenID Provider. OpenPubkey improves on this by providing an optional protocol for the user-identity scenario (interactive browser authentication with the OIDC provider) that removes the OP as a single point of compromise. This protocol independently  authenticates the user via MultiFactor Authentication (MFA) and then cosigns the user's PK Token. We call this the MFA-cosigner. If you desire to remove the OIDC Provider as a single point of compromise, consider requiring the use of an OpenPubkey MFA Cosigner. In the workload identity scenario, this is not supported.

Reviewer: forgot to add the (OP) after "OpenID Provider"

### How does OpenPubkey handle OP (OpenID Provider) public key rollover?

OPs (OpenID Providers) issue ID Tokens by signing them. As required by OpenID Connect, OPs make their public keys avalaible at a JWKS (JSON Web Key Set) URI. Anyone can download the OP's public keys from the JWKS URI and verify an OP's signatures on an ID Token. The location of the JWKS URI is defined in the OPs ["/.well-known/openid-configuration"](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig). OPs rotate the public and signing keys they use for ID Tokens. This rotation frequency is between every two weeks to once a year.

| OpenID Provider | .well-known/openid-configuration | JWKS URI | Key rotation |
| -------------| ------------- | ------------- | ------------- |
| Google | https://accounts.google.com/.well-known/openid-configuration  | https://www.googleapis.com/oauth2/v3/certs |~14 days  |
| Github | https://token.actions.githubusercontent.com/.well-known/openid-configuration | https://www.googleapis.com/oauth2/v3/certs  |~84 days |

> OpenPubkey relies on verifiers being able to check the OP's signature on the ID Token's contained in the PK Token. For many use cases, for instance authenticating access to a server, a user's can just request a new ID Token after the OP rotates their keys. Such use cases do not require that PK Tokens remain verifiable beyond an OP key rotation.

Reviewer: ```suggestion
OpenPubkey relies on verifiers being able to check the OP's signature on the ID Token's contained in the PK Token. For many use cases, such as authenticating access to a server, a user can request a new ID Token after the OP rotates their keys. Such use cases do not require that PK Tokens remain verifiable beyond an OP key rotation.
```

> However in the case in which the PK Token is being used to generate a public signature it is neccessary that verifiers can check the OP's signature on an ID Token even after the OP rotates the keys on the JWKS URI.

Reviewer: ```suggestion
However in the case in which the PK Token is being used to generate a public signature, it is necessary that verifiers can check the OP's signature on an ID Token even after the OP rotates their keys.
```

> Below consider a few of the proposed methods of ensuring the verifiability of signatures after the OP rotates signing keys. This is a non-exhaustive list.

Reviewer 2: ```suggestion
Below are a few of the proposed methods of ensuring the verifiability of signatures after the OP rotates signing keys. This is a non-exhaustive list.
```

#### OP Public Keys via TUF (The Update Framework)

> Docker has proposed including a log of past OP public keys in the signed TUF state which is distributed with Docker Offical Images (DOIs). This seems like a natural fit for the Docker use case as DOIs already depend on the intrigity of TUF.

Reviewer 2: ```suggestion
Docker has proposed including a log of past OP public keys in the signed [TUF](https://theupdateframework.io/) state which is distributed with Docker Offical Images (DOIs). This seems like a natural fit for the Docker use case as DOIs already depend on the integrity of TUF.
```

#### Certificate transparency logs

[Reducing Trust in Automated Certificate Authorities via
Proofs-of-Authentication](https://arxiv.org/pdf/2307.08201.pdf) proposes JWT Ledger, an approach that employs a certificate transparency log and audit mechanism to ensure ID Token's can continue to be verified even after the OP public keys are rotated off of the OP's JWKS URI. It states:

> >To decrease the risk that the JWK Ledger will present false information to users, this ledger is backed by a transparency log.

Reviewer 2: ```suggestion
> To decrease the risk that the JWK Ledger will present false information to users, this ledger is backed by a transparency log.
```

It's safer to have a space between the greater than and the text.

> >"To decrease the risk that the JWK Ledger will present false information to users, this ledger is backed by a transparency log.
> The pace of updates to this log should be relatively low, occurring
> only when the IdP rotates verification keys. Therefore, witnesses
> for the log can verify the current state of the key set on each update; they also check that no entries other than the given key set
> change have been added to the log. Clients can requires a quorum
> of witnesses on the JWK Ledger digest. When a client requests
> the key set for a given timestamp, the ledger serves two entries.
> The client checks that the timestamp of the first entry precedes
> the requested timestamp, that the timestamp of the second entry
> follows the requested timestamp, and that the entries are adjacent
> in the log. This convinces the client that the key set was valid at
> the given time."

Reviewer 2: ```suggestion
>To decrease the risk that the JWK Ledger will present false information to users, this ledger is backed by a transparency log.
> The pace of updates to this log should be relatively low, occurring
> only when the IdP rotates verification keys. Therefore, witnesses
> for the log can verify the current state of the key set on each update; they also check that no entries other than the given key set
> change have been added to the log. Clients can requires a quorum
> of witnesses on the JWK Ledger digest. When a client requests
> the key set for a given timestamp, the ledger serves two entries.
> The client checks that the timestamp of the first entry precedes
> the requested timestamp, that the timestamp of the second entry
> follows the requested timestamp, and that the entries are adjacent
> in the log. This convinces the client that the key set was valid at
> the given time.
```

You don't need the block quote _and_ double quotes.

> This is similar to the approach sketched in [OpenPubkey paper: Appendix C Archival Verification With Certificate Transparency Logs:](https://eprint.iacr.org/2023/296.pdf)

Reviewer 2: ```suggestion
This is similar to the approach sketched in _Appendix C: Archival Verification With Certificate Transparency Logs_ of the [OpenPubkey paper](https://eprint.iacr.org/2023/296.pdf).
```

#### Archival verifiers

> From [OpenPubkey paper: Section 3.5.3 Archival Verification:](https://eprint.iacr.org/2023/296.pdf) proposes archival verifiers which store a log of all past OP public keys. This approach is simple and does not require a transparency log. The main disadvantage is that you can't verify signatures from before their verifier started logging public keys.

Reviewer 2: ```suggestion
_Section 3.5.3: Archival Verification_ of the [OpenPubkey paper](https://eprint.iacr.org/2023/296.pdf) proposes archival verifiers which store a log of all past OP public keys. This approach is simple and does not require a transparency log. The main disadvantage is that you can't verify signatures from before the verifier started logging public keys.
```

# OpenPubkey Reference Implementation

This repo contains the current reference implementation of OpenPubkey. The reference implementation is a work in progress.

## What is OpenPubkey?

OpenPubkey adds user or workload generated public keys to OpenID Connect (OIDC) enabling identities to sign messages or artifacts under their OIDC identity. In essence, OpenPubkey is a protocol for getting OpenID Providers (OPs) to bind identities to public keys.

Verifiers can check that these signatures are valid and associated with the signing OpenID identity. OpenPubkey does not add any new trusted parties beyond what is required for OpenID Connect. It is fully compatible with existing OpenID Providers (Google, Azure/Microsoft, Okta, OneLogin, Keycloak) without any changes to the OpenID Provider.

### ID Tokens and PK Tokens

> OpenPubkey uses ID Tokens issued by OpenID Providers (OPs) to produce PK Tokens. A PK Token consist of the ID Token and the CIC (Client-Instance Claims) where the CIC contains the user's public key and associated metadata. The ID Token contains a hash of the CIC, so that the values in the CIC including the identities public key are cryptographically bound to the ID Token signed by the OP.

Reviewer 2: Should you define the "PK" acronym (public key)?

Author: "PK Token" is just the name of the token, like "ID Token", it is so named because the PK Token contains a public key but I don't think we need qualify it. Elsewhere in the documentation I try to avoid using PK and always say public key instead to avoid confusion with PK=Public Key or Private Key. This is always why I never say Private Key and instead use the terminology Signing Key (SK).

> For workload-identity, the hash of the CIC is stored in the ID Token's `aud` claim following the pattern of github actions workload identity. For user-identity, the hash of the CIC is stored in the ID Token's `nonce` claim.

Reviewer 2: ```suggestion
For workload-identity, the hash of the CIC is stored in the ID Token's `aud` claim following the pattern of GitHub Action's workload identity. For user-identity, the hash of the CIC is stored in the ID Token's `nonce` claim.
```

Reviewer 2: Is it worth adding something to the effect, "Note that the use of the `aud` and `nonce` claims in this manner is completely consistent with the OIDC spec." ?

> “Sigstore [31, 42] is an open source project for signing and verifying software artifacts. Users can sign under their OpenID Connect identity by using the sigstore Fulcio Certificate Authority [43] which uses an immutable log to store a mapping between an OpenID Connect ID Token and a short lived public key enabling parties to attribute signatures to identities. The Fulcio CA (Certificate Authority) is trusted to create this mapping between an ID Token and a public key. Using OpenPubkey this trust can be eliminated as OpenPubkey does not need a trusted party to map ID Tokens to public keys. The Fulcio CA could in turn help OpenPubkey by acting as a public OpenPubkey verifier and OP public key database."

### How does OpenPubkey ensure the nonce claim functions as nonce?

In the user-identity scenario the CIC that contains the user's public key is hashed to the `nonce` claim in the ID Token. As OIDC requires that this field never repeat, OpenPubkey includes a random value, rz, in the CIC. Thus the hash of the CIC is always different and random. This maintains the required properties needed by the `nonce` claim in OIDC.

> ```

Reviewer 2: go language hint?

Author: this is pseudocode, but you are right it does look better with the golang hint

rz = crypto.random()
CIC = {'rz': crypto.random(), 'upk': <publickey>, 'alg': 'EC256'}
IDToken.nonce = SHA3(CIC)
```

### Does OpenPubkey present a privacy leak?

> The PK Tokens used by OpenPubkey contain the claims from the OIDC ID Tokens of the signer, so making them public, necessarily makes those claims public too. This may include elements of the signer’s identity such as the signer’s name or email addresses. It is up to users of OpenPubkey as to whether or not PK Tokens are made public.

Reviewer 2: ```suggestion
The PK Tokens used by OpenPubkey contain the claims from the OIDC ID Tokens of the signer, so making them public necessarily makes those claims public too. This may include elements of the signer’s identity such as the signer’s name or email addresses. It is up to users of OpenPubkey as to whether or not PK Tokens are made public.
```

(remove comma after "so making them public")

> In a public artifact signing scenario, it could be argued that these very claims are the very claims upon which trust in the artifact should be based. However, some OIDC providers may include claims that the signer may wish to keep private. Users of OpenPubkey should consider carefully which OIDC providers to integrate with.

Reviewer 3: ```suggestion
In a public artifact signing scenario, it could be argued that these very claims are the very claims upon which trust in the artifact should be based. However, some OIDC providers may include claims that the signer may wish to keep private. Users of OpenPubkey should consider carefully which OIDC providers to integrate with.
```
```suggestion
In a public artifact signing scenario, it could be argued that these claims are the very claims upon which trust in the artifact should be based. However, some OIDC providers may include claims that the signer may wish to keep private. Users of OpenPubkey should consider carefully which OIDC providers to integrate with.
```

> Most OpenID Providers (OP) allow you to scope the fields in the ID Tokens. For instance Google’s OP is by default scoped to only include userinfo-email claims: name, email address and icon. Given that purpose of OpenPubkey is to enable parties to verify that a particular identity e.g., ethan@bastion.com, produced a particular signature, if you do not want signatures to be associated with OIDC identities, then OpenPubkey may not be a good fit for your use case.

Reviewer 2: ```suggestion
Most OpenID Providers (OP) allow you to scope the fields in the ID Tokens. For instance Google’s OP is by default scoped to only include userinfo-email claims: name, email address and icon. Given that purpose of OpenPubkey is to enable parties to verify that a particular identity, e.g., ethan@bastion.com, produced a particular signature, if you do not want signatures to be associated with OIDC identities, then OpenPubkey may not be a good fit for your use case.
```

(there should be commas before and after "e.g.")

> In other OpenPubkey deployment scenarios, such as those employed by BastionZero, the ID Tokens are not made publicly available.

Reviewer 2: Is it worth mentioning something there is no information leakage in the workflow-identity use case?

Author: I want to address this more completely in the future because it really depends on what is meant by information leakage and the harm it causes. Generally the information in the workload-identity ID Token does not represent any sort of privacy impact but someone can always come up with an absurd counter example.

I'm going to invent an absurd scenario:

Someone could use OpenPubkey for a private repo in which they want to keep the very existence of the repo secret, but they want sign the repo and post the signature and PK Token on a public transparency log. Technically this would be an information leakage. If they want to keep the very existence of their repo secret, I would argue using OpenPubkey with a public transparency log isn't a great fit.

I think this concern requires a longer treatment than I can do right now.

> ### Can the ID Tokens in Pubkeys be replayed against OIDC Resource Providers?

Reviewer 2: Should "Pubkeys" be "OpenPubkey" or "public keys"?

Author: Good catch, this should be "Can the ID Tokens in PK Tokens." In general while I enjoy shortening public key to pubkey, I should always refer to public keys as public keys in official docs as it is the common usuage.

> Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OpenID Provider's signature can be stripped from the ID Token, and a proof of that signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been stripped. However the GQ Signature provides the same assurance that the signature provided.

Reviewer 2: > However the GQ Signature provides the same assurance that the signature provided.

It might be good to further clarify to what signature the second mention of signature in the above sentence refers.

> Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OpenID Provider's signature can be stripped from the ID Token, and a proof of that signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been stripped. However the GQ Signature provides the same assurance that the signature provided.

Reviewer 3: ```suggestion
Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OpenID Provider's signature can be stripped from the ID Token, and a proof of that signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been stripped. However the GQ Signature provides the same assurance that the signature provided.
```
```suggestion
Although not present in the original OpenPubkey paper, GQ signatures have now been integrated so that the OpenID Provider's signature can be stripped from the ID Token, and a proof of that signature published in its place. This prevents the ID Token present in the PK Token from being used against any OIDC resource providers as the original signature has been stripped. However the GQ Signature provides the same assurance that the original OpenID Provider's signature did.
```

- [Guillou and Quisquater, A “Paradoxical” Indentity-Based Signature Scheme Resulting from Zero-Knowledge, Crypto 1988](https://link.springer.com/content/pdf/10.1007/0-387-34799-2_16.pdf)

- [OpenID Connect Core 1.0 incorporating errata set 1](https://openid.net/specs/openid-connect-core-1_0.html)

> - 

Reviewer 2: ?
